### PR TITLE
expression, executor: castStringAsTime returns null when sql_mode is not strict

### PR DESCRIPTION
**Pull request by XuHuaiyu**
<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->
castStringAsTime returns null when sql_mode is not strict.
Before this commit, the following query will raise an error:
``` sql
tidb> set @@sql_mode = '';
Query OK, 0 rows affected (0.00 sec)

tidb> create table t(a datetime);
Query OK, 0 rows affected (0.01 sec)

tidb> insert into t value("1999-12-12");
Query OK, 1 row affected (0.00 sec)

tidb> update t set a = '';
ERROR 1292 (22007): invalid time format: ''
```

### What is changed and how it works?
1. Return warning when sql_mode is not strict in builtinCastStringAsTime.
2. Return null when any error or warning happens in builtinCastStringAsTime.

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Integration test

Code changes

 - Has exported function/method change

Side effects

It's really hard to make all the behavior to be compatiable with MySQL when cast a string as datetime. The results of MySQL are different between implicit cast and explict cast.
e.g. 
``` sql
mysql> set @@sql_mode='';                                                           Query OK, 0 rows affected (0.00 sec)

mysql> select year("20123-12-12");
+---------------------+
| year("20123-12-12") |
+---------------------+
|                NULL | -- The implicit cast result of "20123-12-12" is NULL.
+---------------------+
1 row in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-----------------------------------------+
| Level   | Code | Message                                 |
+---------+------+-----------------------------------------+
| Warning | 1292 | Incorrect datetime value: '20123-12-12' | 
+---------+------+-----------------------------------------+
1 row in set (0.00 sec)

mysql> select year(cast("20123-12-12" as datetime));
+---------------------------------------+
| year(cast("20123-12-12" as datetime)) |
+---------------------------------------+
|                                  NULL | -- The explicit cast result is NULL.
+---------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> update t set a = "20123-12-12";
Query OK, 0 rows affected, 1 warning (0.00 sec)
Rows matched: 1  Changed: 0  Warnings: 1

mysql> select * from t;
+---------------------+
| a                   |
+---------------------+
| 0000-00-00 00:00:00 | -- The implicit cast result of "20123-12-12" is ZeroDatetime.
+---------------------+
1 row in set (0.00 sec)

mysql> update t set a=cast("20123-12-12" as datetime);
Query OK, 1 row affected, 1 warning (0.01 sec)
Rows matched: 1  Changed: 1  Warnings: 1

mysql> select * from t;
+------+
| a    |
+------+
| NULL | -- The explicit cast result is NULL.
+------+
1 row in set (0.00 sec)
```

We return NULL here despite of implict cast or explicit cast. Thus it may break some backward compatibility with MySQL.
e.g.
``` sql
tidb> select unix_timestamp('20102-12-12');
+-------------------------------+
| unix_timestamp('20102-12-12') |
+-------------------------------+
|                          NULL | -- MySQL returns 0 here
+-------------------------------+
1 row in set (0.00 sec)
```
Maybe we need to check the warning as [this comment](https://github.com/pingcap/tidb/pull/7701#discussion_r218076747) mentioned.


2. Some compatibility problems are still not resolved, like the different error message between implicit cast and explicit cast.
e.g.
``` sql
MySQL(root@127.0.0.1:test) > set @@sql_mode='';
Query OK, 0 rows affected (0.00 sec)

MySQL(root@127.0.0.1:test) > insert into t values(cast('' as datetime));                                                                                                                                                                                                                      Query OK, 1 row affected, 1 warning (0.01 sec)

MySQL(root@127.0.0.1:test) > show warnings;
+---------+------+------------------------------+
| Level   | Code | Message                      |
+---------+------+------------------------------+
| Warning | 1292 | Incorrect datetime value: '' | 
+---------+------+------------------------------+
1 row in set (0.00 sec)

MySQL(root@127.0.0.1:test) > insert into t values('');
Query OK, 1 row affected, 1 warning (0.02 sec)

MySQL(root@127.0.0.1:test) > show warnings;
+---------+------+----------------------------------------+
| Level   | Code | Message                                |
+---------+------+----------------------------------------+
| Warning | 1265 | Data truncated for column 'a' at row 1 |
+---------+------+----------------------------------------+
1 row in set (0.00 sec)
```

Related changes

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pingcap/tidb/8516)
<!-- Reviewable:end -->

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
PTAL @tiancaiamao @zz-jason

File: expression/builtin_cast.go [hunk 1]
```diff
@@ -1174,14 +1174,15 @@ func (b *builtinCastStringAsTimeSig) evalTime(row chunk.Row) (res types.Time, is
 	}
 	sc := b.ctx.GetSessionVars().StmtCtx
 	res, err = types.ParseTime(sc, val, b.tp.Tp, b.tp.Decimal)
-	if err != nil {
-		return res, false, errors.Trace(err)
+	hasWarning := err != nil
```
Comment (winoros) [inline expression/builtin_cast.go]: should we name it `hasError`?

File: expression/builtin_cast.go [hunk 2]
```diff
@@ -1174,14 +1174,15 @@ func (b *builtinCastStringAsTimeSig) evalTime(row chunk.Row) (res types.Time, is
 	}
 	sc := b.ctx.GetSessionVars().StmtCtx
 	res, err = types.ParseTime(sc, val, b.tp.Tp, b.tp.Decimal)
-	if err != nil {
-		return res, false, errors.Trace(err)
+	hasErrOrWarning := err != nil
+	if err = handleInvalidTimeError(b.ctx, err); err != nil {
```
Comment (zz-jason) [inline expression/builtin_cast.go]: how about:
```go
if err != nil {
	return types.Time{}, true, handleInvalidTimeError(b.ctx, err)
}

if b.tp.Tp == mysql.TypeDate {
	res.Time = types.FromDate(res.Time.Year(), res.Time.Month(), res.Time.Day(), 0, 0, 0, 0)
}
return res, false, nil
```

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @winoros @tiancaiamao

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -1175,13 +1175,13 @@ func (b *builtinCastStringAsTimeSig) evalTime(row chunk.Row) (res types.Time, is
...
 	if err != nil {
-		return res, false, errors.Trace(err)
+		return types.Time{}, true, handleInvalidTimeError(b.ctx, err)
 	}
 	if b.tp.Tp == mysql.TypeDate {
 		// Truncate hh:mm:ss part if the type is Date.
 		res.Time = types.FromDate(res.Time.Year(), res.Time.Month(), res.Time.Day(), 0, 0, 0, 0)
 	}
-	return res, false, errors.Trace(err)
+	return res, false, err
```
Comment (zz-jason) [inline expression/builtin_cast.go]: `return res, false, nil`

File: expression/builtin_cast.go [hunk 4]
```diff
@@ -1175,13 +1175,13 @@ func (b *builtinCastStringAsTimeSig) evalTime(row chunk.Row) (res types.Time, is
 	sc := b.ctx.GetSessionVars().StmtCtx
 	res, err = types.ParseTime(sc, val, b.tp.Tp, b.tp.Decimal)
 	if err != nil {
-		return res, false, errors.Trace(err)
+		return types.Time{}, true, handleInvalidTimeError(b.ctx, err)
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply expression/builtin_cast.go]: Here the return value is either a NULL or an error. The returned `types.Time{}` is never used.

**Comment by zz-jason** [comment]
@tiancaiamao @winoros @eurekaka PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Comment by zz-jason** [comment]
should we cherry-pick this commit to release-2.1? @XuHuaiyu

**Pull request merged**